File: Software/MenuPrototype/ScrollList.cs
namespace MenuPrototype
{
	partial class Program
	{
		////////////////////////////////////////////////////////////////////
		public class ScrollList : MenuItem
		{
			public ScrollList() : base(string.Empty)
			{
			}

			public override void onSelect()
			{
				if (_subMenuList.Count > 0)
				{
					_subMenuList[_activeSubMenu].onSelect();
				}
			}

			public override bool onKeypressed(int key)
			{
				if (key == btnDOWN)
				{
					_activeSubMenu = adjustWrap(_activeSubMenu, 1, 0, _subMenuList.Count - 1);
				}
				else if (key == btnUP)
				{
					_activeSubMenu = adjustWrap(_activeSubMenu, -1, 0, _subMenuList.Count - 1);
				}
				else
				{
					if (_subMenuList.Count > 0)
					{
						_subMenuList[_activeSubMenu].onKeypressed(key);
					}
				}
				return true;
			}
		}
	}
}

File: Software/Arduino code/OpenAstroTracker/lib/menu/controls/ScrollList.cpp
#include "../../input/LcdButtons.hpp"
#include "../../util/debug.hpp"
#include "../../util/utils.hpp"
#include "ScrollList.hpp"

ScrollList::ScrollList() : MenuItem("S","SCRL")
{
}

void ScrollList::onSelect()
{
    if (_subMenuList->count() > 0)
    {
        _subMenuList->getItem(_activeSubMenu)->onSelect();
    }
}

bool ScrollList::onKeypressed(int key)
{
    if (key == btnDOWN)
    {
        //LOGV1(255, "SCROLL: Down pressed");
        _activeSubMenu = adjustWrap(_activeSubMenu, 1, 0, _subMenuList->count() - 1);
    }
    else if (key == btnUP)
    {
        //LOGV1(255, "SCROLL: Up pressed");
        _activeSubMenu = adjustWrap(_activeSubMenu, -1, 0, _subMenuList->count() - 1);
    }
    else 
    {
        if (_subMenuList->count() > 0)
        {
            return _subMenuList->getItem(_activeSubMenu)->onKeypressed(key);
        }
    }
    return true;
}
